File: app/src/main/java/com/dominikgaller/alpinebooking/booking/core/domain/TourBooking.java
package com.dominikgaller.alpinebooking.booking.core.domain;

import com.dominikgaller.alpinebooking.booking.core.domain.event.DomainEvent;
import com.dominikgaller.alpinebooking.booking.core.domain.event.TourBookingCancelled;
import com.dominikgaller.alpinebooking.booking.core.domain.event.TourBookingConfirmed;
import com.dominikgaller.alpinebooking.booking.core.domain.event.TourBookingRequested;
import com.dominikgaller.alpinebooking.booking.core.domain.exception.CapacityExceededException;
import com.dominikgaller.alpinebooking.booking.core.domain.exception.InvalidBookingRequestException;
import com.dominikgaller.alpinebooking.booking.core.domain.exception.InvalidBookingStateException;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Aggregate root representing a reservation for a guided alpine tour.
 *
 * <p>State changes are performed via named methods ({@link #request}, {@link #confirm},
 * {@link #cancel}). Each method records the resulting domain event internally and exposes
 * events via {@link #pullDomainEvents()}.
 *
 * <p>Reconstitution from persistence uses {@link #reconstitute} — no invariants are
 * re-checked when loading an already-valid past fact.
 *
 * <p>Framework-free: no Spring, no JPA, no IO.
 *
 * <p>SDD: See {@code documentation/domain/aggregate-tour-booking.spec.md}.
 */
public class TourBooking {

    private final BookingId bookingId;
    private final TourId tourId;
    private final TourDate tourDate;
    private final ParticipantCount participantCount;
    private final AvailableCapacity availableCapacity;
    private final ParticipantContact contact;
    private TourBookingStatus status;

    private final List<DomainEvent> domainEvents = new ArrayList<>();

    private TourBooking(
            final BookingId bookingId,
            final TourId tourId,
            final TourDate tourDate,
            final ParticipantCount participantCount,
            final AvailableCapacity availableCapacity,
            final ParticipantContact contact,
            final TourBookingStatus status) {
        this.bookingId = bookingId;
        this.tourId = tourId;
        this.tourDate = tourDate;
        this.participantCount = participantCount;
        this.availableCapacity = availableCapacity;
        this.contact = contact;
        this.status = status;
    }

    /**
     * Creates a new booking in {@code REQUESTED} state.
     *
     * @param bookingId         unique identity; must not be null
     * @param tourId            tour reference; must not be null
     * @param tourDate          scheduled date; must be in the future relative to {@code now}
     * @param participantCount  number of participants; must be >= 1
     * @param availableCapacity open spots at booking time; must be >= participantCount
     * @param contact           contact person for the booking; must not be null
     * @param now               current time used to validate the tour date
     * @return a valid {@link TourBooking} in {@code REQUESTED} state
     * @throws InvalidBookingRequestException if {@code tourDate} is not in the future
     * @throws CapacityExceededException      if {@code participantCount} exceeds {@code availableCapacity}
     */
    public static TourBooking request(
            final BookingId bookingId,
            final TourId tourId,
            final TourDate tourDate,
            final ParticipantCount participantCount,
            final AvailableCapacity availableCapacity,
            final ParticipantContact contact,
            final Instant now) {

        if (!tourDate.isInFuture(now)) {
            throw new InvalidBookingRequestException(
                    "Tour date must be in the future, was: " + tourDate.value());
        }
        if (participantCount.value() > availableCapacity.value()) {
            throw new CapacityExceededException(
                    participantCount.value(), availableCapacity.value());
        }

        final TourBooking booking = new TourBooking(
                bookingId, tourId, tourDate, participantCount,
                availableCapacity, contact, TourBookingStatus.REQUESTED);

        booking.domainEvents.add(new TourBookingRequested(
                bookingId, tourId, tourDate, participantCount, now));

        return booking;
    }

    /**
     * Reconstitutes a {@link TourBooking} from its persisted state.
     *
     * <p>No creation-time invariants are enforced — the data is assumed to have been
     * valid when first written. Called exclusively by the persistence mapper.
     *
     * @return a {@link TourBooking} reflecting the stored state, with no pending events
     */
    public static TourBooking reconstitute(
            final BookingId bookingId,
            final TourId tourId,
            final TourDate tourDate,
            final ParticipantCount participantCount,
            final AvailableCapacity availableCapacity,
            final ParticipantContact contact,
            final TourBookingStatus status) {
        return new TourBooking(
                bookingId, tourId, tourDate, participantCount,
                availableCapacity, contact, status);
    }

    /**
     * Transitions the booking from {@code REQUESTED} to {@code CONFIRMED}.
     *
     * @throws InvalidBookingStateException if the current state is not {@code REQUESTED}
     */
    public void confirm(final Instant now) {
        if (status != TourBookingStatus.REQUESTED) {
            throw new InvalidBookingStateException(status);
        }
        status = TourBookingStatus.CONFIRMED;
        domainEvents.add(new TourBookingConfirmed(bookingId, now));
    }

    /**
     * Transitions the booking from {@code REQUESTED} or {@code CONFIRMED} to {@code CANCELLED}.
     *
     * @throws InvalidBookingStateException if the current state is neither {@code REQUESTED}
     *                                      nor {@code CONFIRMED}
     */
    public void cancel(final Instant now) {
        if (status != TourBookingStatus.REQUESTED && status != TourBookingStatus.CONFIRMED) {
            throw new InvalidBookingStateException(status);
        }
        status = TourBookingStatus.CANCELLED;
        domainEvents.add(new TourBookingCancelled(bookingId, now));
    }

    /**
     * Returns and clears all recorded domain events.
     *
     * <p>Calling this method twice returns an empty list on the second call.
     *
     * @return unmodifiable snapshot of pending events
     */
    public List<DomainEvent> pullDomainEvents() {
        final List<DomainEvent> snapshot = Collections.unmodifiableList(
                new ArrayList<>(domainEvents));
        domainEvents.clear();
        return snapshot;
    }

    public BookingId bookingId() {
        return bookingId;
    }

    public TourId tourId() {
        return tourId;
    }

    public TourDate tourDate() {
        return tourDate;
    }

    public ParticipantCount participantCount() {
        return participantCount;
    }

    public AvailableCapacity availableCapacity() {
        return availableCapacity;
    }

    public ParticipantContact contact() {
        return contact;
    }

    public TourBookingStatus status() {
        return status;
    }
}

File: app/src/main/kotlin/com/example/service/booking/core/domain/tourbooking/TourBooking.kt
package com.example.service.booking.core.domain.tourbooking

import com.example.service.booking.core.domain.tourbooking.event.BookingActivated
import com.example.service.booking.core.domain.tourbooking.event.TourBookingConfirmed
import com.example.service.booking.core.domain.tourbooking.event.TourBookingRequested
import com.example.service.booking.core.domain.tourbooking.exception.CapacityExceededException
import com.example.service.booking.core.domain.tourbooking.exception.InvalidBookingRequestException
import com.example.service.booking.core.domain.tourbooking.exception.InvalidBookingStateException
import com.example.service.shared.domain.TourId
import com.example.service.shared.domain.event.DomainEvent
import java.time.Instant

/**
 * Aggregate root representing a reservation for a guided tour.
 *
 * State changes happen through named transition methods — [confirm], [markActive] — each
 * of which enforces the invariants of that transition and records the resulting domain
 * event. Events are handed off through [pullDomainEvents].
 *
 * Deliberately **not** a `data class`: structural equality is wrong for an entity, and a
 * generated `copy()` would hand every caller a way around these transition methods
 * (`coding-style.definition.md` § 2.1). Mutable state uses `private set` for the same
 * reason (§ 5.1).
 *
 * Reconstitution from persistence goes through [reconstitute], which re-checks no
 * invariants: the row was valid when it was written, and re-validating a past fact means
 * a schema change can make history unreadable.
 *
 * Framework-free: no Spring, no JPA, no IO.
 *
 * `LongParameterList` is suppressed for the whole class. An aggregate assembles its
 * entire state in one constructor and its factories mirror that, so the count is a
 * property of the state, not of a behaviour with too many knobs — which is what the
 * rule is for, and which its default threshold still catches everywhere else. The two
 * ways to satisfy it here would be a parameter object nothing else uses, or setters,
 * and setters on an aggregate are forbidden outright (`coding-style.definition.md`
 * § 5.1).
 *
 * SDD: see `documentation/domain/aggregate-tour-booking.spec.md`.
 */
@Suppress("LongParameterList")
class TourBooking private constructor(
    val bookingId: BookingId,
    val tourId: TourId,
    val tourDate: TourDate,
    participantCount: ParticipantCount,
    availableCapacity: AvailableCapacity,
    val contact: ParticipantContact,
    status: TourBookingStatus,
) {
    var participantCount: ParticipantCount = participantCount
        private set

    var availableCapacity: AvailableCapacity = availableCapacity
        private set

    var status: TourBookingStatus = status
        private set

    private val domainEvents = mutableListOf<DomainEvent>()

    /**
     * Transitions a requested booking to `CONFIRMED` (UC02).
     *
     * @param now the moment of confirmation, supplied by the driver from `ClockPort`
     * @throws InvalidBookingStateException if the booking is not `REQUESTED`
     */
    fun confirm(now: Instant) {
        if (status != TourBookingStatus.REQUESTED) {
            throw InvalidBookingStateException(status)
        }
        status = TourBookingStatus.CONFIRMED
        domainEvents += TourBookingConfirmed(bookingId, now)
    }

    /**
     * Transitions a confirmed booking to `ACTIVE` because its tour started (UC06).
     *
     * Idempotent: activating an already-active booking is a no-op that records no event.
     * That is not politeness, it is required — the caller is an `AFTER_COMMIT` listener,
     * and a redelivered event must not fail the second time.
     *
     * `REQUESTED -> ACTIVE` is deliberately rejected. Tolerating it would let a booking
     * nobody confirmed be carried along by a tour starting.
     *
     * @param startedAt the moment tour execution began
     * @param guideTourId opaque correlation id owned by the `guide` context; may be null
     * @throws InvalidBookingStateException if the booking is neither `CONFIRMED` nor
     *   already `ACTIVE`
     */
    fun markActive(
        startedAt: Instant,
        guideTourId: String?,
    ) {
        if (status == TourBookingStatus.ACTIVE) {
            return
        }
        if (status != TourBookingStatus.CONFIRMED) {
            throw InvalidBookingStateException(status)
        }
        status = TourBookingStatus.ACTIVE
        domainEvents += BookingActivated(bookingId, startedAt, guideTourId)
    }

    /**
     * Returns and clears the recorded domain events.
     *
     * A second call returns an empty list. The snapshot is a copy, so a caller iterating
     * it cannot be surprised by a later transition.
     */
    fun pullDomainEvents(): List<DomainEvent> {
        val snapshot = domainEvents.toList()
        domainEvents.clear()
        return snapshot
    }

    companion object {
        /**
         * Creates a booking in `REQUESTED` state (UC01).
         *
         * @throws InvalidBookingRequestException if [tourDate] is not in the future
         * @throws CapacityExceededException if [participantCount] exceeds
         *   [availableCapacity]
         */
        fun request(
            bookingId: BookingId,
            tourId: TourId,
            tourDate: TourDate,
            participantCount: ParticipantCount,
            availableCapacity: AvailableCapacity,
            contact: ParticipantContact,
            now: Instant,
        ): TourBooking {
            if (!tourDate.isInFuture(now)) {
                throw InvalidBookingRequestException(
                    "Tour date must be in the future, was: ${tourDate.value}",
                )
            }
            if (participantCount.value > availableCapacity.value) {
                throw CapacityExceededException(participantCount.value, availableCapacity.value)
            }

            return TourBooking(
                bookingId = bookingId,
                tourId = tourId,
                tourDate = tourDate,
                participantCount = participantCount,
                availableCapacity = availableCapacity,
                contact = contact,
                status = TourBookingStatus.REQUESTED,
            ).also {
                it.domainEvents += TourBookingRequested(bookingId, tourId, tourDate, participantCount, now)
            }
        }

        /**
         * Rebuilds a booking from its persisted state, with no pending events.
         *
         * Called by the persistence mapper. `ClassRoleRulesTest` enforces that: a caller
         * outside `..outbound.persistence..` reaching for this method is skipping the
         * invariants that [request] enforces.
         */
        fun reconstitute(
            bookingId: BookingId,
            tourId: TourId,
            tourDate: TourDate,
            participantCount: ParticipantCount,
            availableCapacity: AvailableCapacity,
            contact: ParticipantContact,
            status: TourBookingStatus,
        ): TourBooking =
            TourBooking(
                bookingId = bookingId,
                tourId = tourId,
                tourDate = tourDate,
                participantCount = participantCount,
                availableCapacity = availableCapacity,
                contact = contact,
                status = status,
            )
    }
}
